### ISSUE TITLE
bash: - : invalid option

### ISSUE THREAD

**Issue by 3togo** [mxe/mxe#2659]
How to get rid of the following warning?

eli@lzx-acer:/mnt/data/git/jos/mxe$ make 
bash: - : invalid option
Usage:	bash [GNU long option] [option] ...
	bash [GNU long option] [option] script-file ...
GNU long options:
	--debug
	--debugger
	--dump-po-strings
	--dump-strings
	--help
	--init-file
	--login
	--noediting
	--noprofile
	--norc
	--posix
	--pretty-print
	--rcfile
	--restricted
	--verbose
	--version
Shell options:
	-ilrsD or -c command or -O shopt_option		(invocation only)
	-abefhkmnptuvxBCHP or -o option
[build]       mxe-conf                i686-w64-mingw32.static

**Comment by DerOrfa** [comment]
Had the same issue. Changing `python` to `python2` fixed it.
There is no `python` in Ubuntu (only python2 and python3). 
So `which python` returns an empty string into the variable `PYTHON` so any command  `$(shell $(PYTHON) -c ...` (e.g. the line just below) ends up being `bash -c ..`. 
Which bash knows nothing about.

**Comment by Light-Wizzard** [comment]
Adding alias for python to your profile fixes this issue.
`alias python=python3`
or
`alias python=python2`

**Comment by DerOrfa** [comment]
Agreed. Its an OS problem.
But maybe check the result of https://github.com/mxe/mxe/blob/master/Makefile#L42. 
It might help prevent users from getting weird error messages if there is no python executable (for whatever reason).

**Comment by Light-Wizzard** [comment]
I would just write a bash file to run the above, messing with a make file is not a very good option for a few reasons, you could check the value of which python, but this should have been done before running make, it is not just the make file that is confused about where python is, and this is OS-specific or distro-specific, mostly only affecting Ubuntu and Debian, I would have never noticed if I did not run this on an Ubuntu KVM, I run Arch Linux Manjaro, so the problem is deeper, the above bash scripts are meant to run once, and not every time, I run them in my installation bash script, if you run the check for python2, which python might return a path, it might not, depending on what version, I have seen -f /usr/bin/python2  return true while which python returns a blank, and you have the case where python is not installed, all these things should be checked before running make, so what I would suggest is something like this, I have not tested the code, it is just an example, but should work, I have been writing bash commands for over 30 years, but I am also over 60 with poor eye sight, but my point is this is not a Make issue, it assumes the machine is properly setup, and which python drawing a blank is not setup right, the concept of python, python2 and python3, is sad to begin with, python should be backward compatible and have no need for this, but it is not forward compatible, and why we have this concept, so we can use older versions, and in this case,  python2 seems to fail, and python3 did no better, so I have to find out what version of python it requires, because at this point, my guess is python 1, and if that is the case, this script will not help, you need to write a script to ensure that version is installed.

```bash
# Check to see if bash_aliases exist, create if it not.
if [ ! -f ~/.bash_aliases ]; then 
    touch ~/.bash_aliases; 
fi; 
# Track if Python exists
declare -i IS_PYTHON; IS_PYTHON=0;
# check python first
if [ "$(which python)" != "" ]; then
    IS_PYTHON=1; 
fi
# Check python2 
if [ "$IS_PYTHON" -eq 0 ]; then 
    if [ -f /usr/bin/python2 ]; then 
        if [ "$(which python)" != "" ]; then
            sudo ln -s /usr/bin/python2 /usr/bin/python; 
            if [ "$(which python)" != "" ]; then
                echo "alias python=python2;" >> ~/.bash_aliases; alias python=python2;
                IS_PYTHON=1;    
            fi
        fi
    fi
fi
# Check python3
if [ "$IS_PYTHON" -eq 0 ]; then 
    if [ -f /usr/bin/python3 ]; then 
        if [ "$(which python)" != "" ]; then
            sudo ln -s /usr/bin/python3 /usr/bin/python; 
            if [ "$(which python)" != "" ]; then
                echo "alias python=python3;" >> ~/.bash_aliases; alias python=python3;
                IS_PYTHON=1;    
            fi
        fi
    fi
fi
# unless we have python4 we stop here
if [ "$IS_PYTHON" -eq 0 ]; then 
    # python is not installed 
    # check distro to give correct install command
    echo "install Python";
fi

```

If you want to check for a specific version of python

```bash
declare -i PYTHON_VERSION_REQUIRE; PYTHON_VERSION_REQUIRE=1;

if [ "$(python -c 'import sys; print(".".join(map(str, sys.version_info[:1])))')" -ne "$PYTHON_VERSION_REQUIRE" ]; then 
    # install or link to the the right version
fi

if [ "$(python -c 'import sys; print(".".join(map(str, sys.version_info[:1])))')" -eq 1 ]; then 
    # do something
elif [ "$(python -c 'import sys; print(".".join(map(str, sys.version_info[:1])))')" -eq 2 ]; then 
    # do something
elif [ "$(python -c 'import sys; print(".".join(map(str, sys.version_info[:1])))')" -eq 3 ]; then 
    # do something
fi

```

Whatever you do to fix this needs to be done before configure or "make", linking will not work otherwise.

Flesh.

**Comment by rkdover** [comment]
I also had this option but the makefile didn't fail until the second mention of python with a `python is not recognized as a command` error. Fixed it by creating a link to it with ´ln /usr/bin/python3 /usr/bin/python´ and it built just fine. Alias didn't fix the issue for me.

Shouldn't python (and libssl-dev which is needed later on) be part of the makefile dependency check?

(I'm building on a docker image of debian)

**Comment by Light-Wizzard** [comment]
You should have said that from the beginning, your issues are with docker, and how a container works, it has a base image, you can change settings in the docker image itself, you have to change them at run time, I am not a docker person, maybe this link will help, maybe its a different subject https://docs.docker.com/language/python/build-images/ but I do know if you want a docker image to work out of the box with python, you have to make changes to the images, or use python2 or python3 instead of python, or soft link them as you have, the alias is to alias python to use python2 or 3 instead, and only works at the command level, it does not make links to the file if that is what you mean by does not work, if you can run the command python2 or pthon3, then alias should work, but if those do not run, neither will the alias.

You should look for a docker image that is set up for the version of python you need, the problem with asking make to fix a problem with the way the system is set up would be opening up a can of worms, it takes for granted you have the machine set up for python, sure it could do a check, but asking it to pick what version of python you want to use is beyond a simple check, it is a can of worms, because it has to assume you meant 2 or 3, and why "make" is the wrong place to make this check, it should be done before you decide to compile anything, you have to make sure your system is set up for the right version of python, you do not leave it up to make, to make that decision for you, and that is what you are asking for, a can of worms, it has no idea what version you want, maybe it should have a way, maybe have built-in regression testing, see what I mean, how is make going to know what version you need to pass all the test, and why I call it can of worms.

**Comment by rkdover** [comment]
> You should have said that from the beginning, your issues are with docker, and how a container works, it has a base image, you can change settings in the docker image itself, you have to change them at run time, I am not a docker person, maybe this link will help, maybe its a different subject https://docs.docker.com/language/python/build-images/ but I do know if you want a docker image to work out of the box with python, you have to make changes to the images, or use python2 or python3 instead of python, or soft link them as you have, the alias is to alias python to use python2 or 3 instead, and only works at the command level, it does not make links to the file if that is what you mean by does not work, if you can run the command python2 or pthon3, then alias should work, but if those do not run, neither will the alias.
> 
> 
> 
> You should look for a docker image that is set up for the version of python you need, the problem with asking make to fix a problem with the way the system is set up would be opening up a can of worms, it takes for granted you have the machine set up for python, sure it could do a check, but asking it to pick what version of python you want to use is beyond a simple check, it is a can of worms, because it has to assume you meant 2 or 3, and why "make" is the wrong place to make this check, it should be done before you decide to compile anything, you have to make sure your system is set up for the right version of python, you do not leave it up to make, to make that decision for you, and that is what you are asking for, a can of worms, it has no idea what version you want, maybe it should have a way, maybe have built-in regression testing, see what I mean, how is make going to know what version you need to pass all the test, and why I call it can of worms.
> 
> 

I didn't open the issue, so I couldn't have mentioned it earlier. 😄 

Besides, the makefile doesn't check for any version of python to my knowledge. If that's not me missing something, I would say that's a makefile issue.

**Comment by Light-Wizzard** [comment]
I agree that make does not make python checks, what you need to do is be more specific about what type of checks you want it to do, then figure out how to implement them on all platforms, I am sure implemented a linkage check to ensure python is linked to python 2 or 3, but I do not see a way to tell it which one you want to use, and that is the problem with having make check python, it really has no idea what version you want, so I am at a loss to what types of checks it can do, maybe a feature request to add a switch to tell it which version of python to use, that is more specific than just a generic check, I run into this issue all the time on AppVeyor, and wished cmake or make could resolve this issue given an argument as to what version you want, but I always look for a way to implement it, or at least have a vague concept of how this can be accomplished.

This problem goes back to how python is versioned, but given a switch to tell it which version, there is no reason it could not make the linkage to the correct version if it does not exist, I just do this in my bash scripts, because I cannot change the image that AppVeyor uses, but the biggest issue is the all platforms issues, even the location of Python changes on which distribution of Linux you use, not to mention Windows or Mac, and a Docker image is another issue, there are way too many images that have different python configurations, grant you on my PI and Debian based systems I know I can count on where python is installed by default, but whereis might give you more than one option, so you would have to base which to chose on some switch you can change at compile-time, I end up with a different script to deal with this on every system I work on, Python has no set standard across all platforms, but if you could work out a system that works, and some code to demo it, the feature would be great, my point is be very specific about what you are asking for, and be more specific about how it can be implemented across all platforms, I know I could work out the logic and test it on KVM machines, because that is what whoever takes on a feature like this will have to do, so maybe take this discussion in a direction that includes some code so people can test it, even a bash script would be something that can be converted into code.

**Comment by LQss11** [comment]
It is due to the script getting changed into Windows format. To fix it open the script file in Notepad++ , click on ‘Edit’–>‘EOL Conversion’–>‘Unix (LF)’ and save the file. [reference](https://community.grafana.com/t/invalid-option-running-docker-image/40176/2)

**Comment by Yasuomvp** [comment]
> It is due to the script getting changed into Windows format. To fix it open the script file in Notepad++ , click on ‘Edit’–>‘EOL Conversion’–>‘Unix (LF)’ and save the file. [reference](https://community.grafana.com/t/invalid-option-running-docker-image/40176/2)

Correct!   I've been tormented by this problem for 3 days....   QAQ
